### PR TITLE
Added coding standards for model implementations as a custom context for greptile

### PR DESCRIPTION
**Pull request by CharlelieLrt**
# PhysicsNeMo Pull Request

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: CODING_STANDARDS/MODELS_IMPLEMENTATION.md [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+complete models that are applicable to a specific domain or specific data
+modality.
+
+---
+
+### MOD-001: Use proper class inheritance for all models
+
+**Description:**
+All model classes must inherit from `physicsnemo.Module`. Subclasses of
+`torch.nn.Module` are not allowed.
```
Comment (coreyjadams) [inline CODING_STANDARDS/MODELS_IMPLEMENTATION.md]: We should maybe say _direct_ subclasses are not allowed, but physicsnemo.Module is, itself, a subclass of torch.nn.Module.

File: CODING_STANDARDS/MODELS_IMPLEMENTATION.md [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+Ensure proper initialization of parent classes using `super().__init__()`. Pass
+the `meta` argument to the `super().__init__()` call if appropriate, otherwise
+set it manually with `self.meta = meta`.
+
+**Rationale:**
+Ensures invariants and functionality of the `physicsnemo.Module` class for all
+models.
+
```
Comment (coreyjadams) [inline CODING_STANDARDS/MODELS_IMPLEMENTATION.md:171]: I think we should expand on this.  The core motivations for this requirement should be listed here, for users who want to know:

- Enable better serialization of models that use sub-pnm-modules.

...Is there another reason?

File: CODING_STANDARDS/MODELS_IMPLEMENTATION.md [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+**Example:**
+
+```python
+from physicsnemo.models import Module
```
Comment (coreyjadams) [inline CODING_STANDARDS/MODELS_IMPLEMENTATION.md]: This should probably be `physicsnemo.core`

File: CODING_STANDARDS/MODELS_IMPLEMENTATION.md [hunk 4]
```diff
@@ -0,0 +1,484 @@
...
+  folders directly, and backward compatibility is guaranteed. Another exception
+  is when the model class is highly specific to a single example. In this case,
+  it may be acceptable to place it in a module specific to the example code,
+  such as for example `examples/<example_name>/utils/nn.py`.
+
+- Stage 2: Production. After staying in stage 1 for a sufficient amount of time
+  (typically at least 1 release cycle), the model class is promoted to stage 2.
+  It is then moved to the `physicsnemo/nn` or `physicsnemo/models` folders,
+  based on the rule `MOD-000`. During this stage, backward compatibility is
+  guaranteed.
```
Comment (coreyjadams) [inline CODING_STANDARDS/MODELS_IMPLEMENTATION.md:218]: We should consider what "production" models look like re:checkpoints.  Are they meant to be compatible with pre-trained instances on HF or NGC?  Do we need any versioning?

File: CODING_STANDARDS/MODELS_IMPLEMENTATION.md [hunk 5]
```diff
@@ -0,0 +1,484 @@
...
+  are not recognized by our Sphinx documentation and are prohibited.
+
+- All methods should be documented with a docstring, with at least a `Parameters`
+  section and a `Returns` section. Other sections such as `Notes`, `Examples`,
+  or `..important::` or `..code-block:: python` are possible. Other sections
+  are not recognized by our Sphinx documentation and are prohibited.
+
+- All tensors should be documented with their shape, using LaTeX math notation such
+  as :math:`(N, C, H_{in}, W_{in})` (there is flexibility for naming the
+  dimensions, but te math format should be enforced). Our documentation is
```
Comment (coreyjadams) [inline CODING_STANDARDS/MODELS_IMPLEMENTATION.md]: te -> the

File: CODING_STANDARDS/MODELS_IMPLEMENTATION.md [hunk 6]
```diff
@@ -0,0 +1,484 @@
...
+        :math:`(B, D_{latent})`. If ``True``, returns tuple of
+        (latent, hidden) where hidden has shape :math:`(B, D_{hidden})`.
+
+    Examples
+    # Rule: Examples section is allowed and helpful
+    --------
+    >>> model = SimpleEncoder(input_dim=784, latent_dim=64)
+    >>> x = torch.randn(32, 784)
+    >>> latent = model(x)
+    >>> latent.shape
```
Comment (coreyjadams) [inline CODING_STANDARDS/MODELS_IMPLEMENTATION.md:465]: We should have a comment in the guide about this.  It's excellent to have; we could to note to users that code examples must run successfully since our CI checks it.

File: CODING_STANDARDS/MODELS_IMPLEMENTATION.md (hunk 1 above)
Comment (CharlelieLrt) [reply CODING_STANDARDS/MODELS_IMPLEMENTATION.md]: Edited

File: CODING_STANDARDS/MODELS_IMPLEMENTATION.md (hunk 2 above)
Comment (CharlelieLrt) [reply CODING_STANDARDS/MODELS_IMPLEMENTATION.md:171]: Added more details: serialization and checkpointing as you said, versioning system for backward compatibility, and can be registered in the registry (not possible with `torch.nn.Module`)

File: CODING_STANDARDS/MODELS_IMPLEMENTATION.md (hunk 3 above)
Comment (CharlelieLrt) [reply CODING_STANDARDS/MODELS_IMPLEMENTATION.md]: Or just `from physicsnemo import Module`? Are we planning to expose `Module` in the top-level package like it is currently the case?

File: CODING_STANDARDS/MODELS_IMPLEMENTATION.md (hunk 6 above)
Comment (CharlelieLrt) [reply CODING_STANDARDS/MODELS_IMPLEMENTATION.md:465]: Do you mean all models should have a runnable `Examples` section? For some we only have `.. code-block::`, would that be fine or do you think Examples is needed?

File: CODING_STANDARDS/MODELS_IMPLEMENTATION.md (hunk 4 above)
Comment (CharlelieLrt) [reply CODING_STANDARDS/MODELS_IMPLEMENTATION.md:218]: Good point. Actually the model versioning system is a key feature of `physicsnemo.Module`, and I think it deserves its own rule. I'll add one

**Comment by CharlelieLrt** [comment]
@coreyjadams following up on these 2 things.

> Production Testing structure. What tests are minimally required, what tests are encouraged? We should note that models can not move out of experimental without complete test coverage.

Right, I was planning on adding a rule for that.

> Requirements. For models that introduce extra requirements, we should prescribe a method to note and track that. TBD on managing requirements in general so I'm not sure we need a solution now, but we should update it here when we get there.

What type of mechanism are you thinking of here? For example if something use apex and its not available we use a try/catch:

```python
try:
    import apex
except ImportError:
    OptionalDependencyFailure("apex")
    apex = None
```

It's the pattern used in E2S for optional dependencies. What about cases where there is a fallback (e.g. many of apex layers can just fallback to torch layers in case apex is not available)?

File: CODING_STANDARDS/MODELS_IMPLEMENTATION.md (hunk 6 above)
Comment (coreyjadams) [reply CODING_STANDARDS/MODELS_IMPLEMENTATION.md:465]: I don't think every model needs a runnable example, what I meant was that this section - if present - will get tested by CI so it's worth a note in the guidelines about it.  Something like ...

```
Your docstring is encouraged to have an example section to demonstrate basic construction and usage.  Note that our CI system will automatically test these sections for correctness, when present.
```

I wasn't trying to force anything complicated here :)

**Comment by coreyjadams** [comment]
re:requirements ...

I'm not sure what the right solution is.  Ideally, I want a system in physicsnemo that can do the following for us:
- For missing required dependencies, fail with an informative message.
- For missing optional dependencies, warn with an informative message without overwhelming.
- For models, track with CI what is used in the model vs. what is stated in the requirements.  If a user or developer adds a new dependency, but doesn't update the requirements, CI should reject / fix it.
- Ideally, avoid too many try/except blocks that then have conditional paths laters.  Can't always be avoided of course.

I just want to get to a state where we have consistent requirements to make installations easier for target models, if that makes sense, without bogging us all down in import paperwork more or less.  I haven't actually got a plan for implementation, this is just daydreaming at the moment.

File: CODING_STANDARDS/MODELS_IMPLEMENTATION.md (hunk 4 above)
Comment (CharlelieLrt) [reply CODING_STANDARDS/MODELS_IMPLEMENTATION.md:218]: Added rule for versioning models.

File: CODING_STANDARDS/MODELS_IMPLEMENTATION.md (hunk 3 above)
Comment (CharlelieLrt) [reply CODING_STANDARDS/MODELS_IMPLEMENTATION.md]: Changed to `from physicsnemo import Module`
